### PR TITLE
refactor and improve xline server

### PR DESCRIPTION
**Pull request by markcty**
1. Watcher should be an inner struct in the storage module. It should not appear in the server module. WatchServer only needs to maintain watch_id, no need to maintain watcher.
2. Some renaming
3. Use a RwLock to store watchers so that different watchers can send concurrently
4. Unnecessary Arc<Inner>: Watcher has an inner of type Arc<WatcherInner>. When Watcher needs to be shared, we previously Clone on Watcher. Just use Arc<Watcher> is both clearer and simpler.
5. Improve filter events performance: previously events are filtered by key_range twice, now only once
6. Watch id should be assigned from 1 instead of 0 because 0 means auto-generated, which might be confusing to the client
7. Reduce code duplication
8. Use atomici64 for the revision number.
9. Fix in auth backend concurrent bugs due to the revision number

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Refactor and improve KV watch

* what changes does this pull request make?
Refactors

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
No.

### CODE REVIEW

File: xline/src/storage/authstore/backend.rs [hunk 1]
```diff
@@ -714,143 +715,131 @@ impl AuthStoreBackend {
...
         self.put(AUTH_ENABLE_KEY.to_vec(), vec![0], revision, 0);
         self.enabled.store(false, AtomicOrdering::Release);
-        true
     }
 
     /// Sync `AuthStatusRequest` and return whether authstore is changed.
-    fn sync_auth_status_request(_req: &AuthStatusRequest) -> bool {
-        false
-    }
+    fn sync_auth_status_request(_req: &AuthStatusRequest) {}
```
Comment (Phoenix500526) [inline xline/src/storage/authstore/backend.rs:817]: What's the point of existence of this function? It has neither a return value nor a function body

File: xline/tests/auth_test.rs [hunk 2]
```diff
@@ -186,12 +186,12 @@ async fn test_auth_wrong_password() -> Result<(), Box<dyn Error>> {
     .await;
     assert!(result.is_err());
 
-    let result = etcd_client::Client::connect(
-        vec![cluster.addrs()["server0"].to_string()],
-        Some(ConnectOptions::new().with_user("root", "123")),
-    )
-    .await;
-    assert!(result.is_ok());
+    // let result = etcd_client::Client::connect(
```
[the remark this answers is not in the document]
Comment (markcty) [reply xline/tests/auth_test.rs]: My bad. Forgot to remove the comment after debugging

File: xline/src/storage/authstore/backend.rs (hunk 1 above)
Comment (markcty) [reply xline/src/storage/authstore/backend.rs:817]: I guess @themanforfree wants to keep it so that all the sync_... functions have the same form.

**Review by themanforfree — APPROVED** [review]

File: xline/src/storage/auth_store/backend.rs [hunk 3]
```diff
@@ -947,20 +929,20 @@ impl AuthStoreBackend {
...
-            Err(_) => return false,
+            Err(_) => return,
         };
         let idx = match user.roles.binary_search(&req.role) {
             Ok(idx) => idx,
-            Err(_) => return false,
+            Err(_) => return,
         };
         let _ignore = user.roles.remove(idx);
+        let revision = self.revision.get();
```
Comment (themanforfree) [inline xline/src/storage/auth_store/backend.rs:945]: this should be revision.next()

Comment (markcty) [reply xline/src/storage/auth_store/backend.rs]: Fixed.

**Review by themanforfree — APPROVED** [review]

**Review by Phoenix500526 — APPROVED** [review]

**Pull request merged**